lib/sgtn-client/api/translation.rb

```diff
@@ -67,97 +69,42 @@ def self.getTranslation(component, key, locale)
-        # TODO: if an error occurs when requesting a bundle, need to avoid more requests
-      elsif CacheUtil.is_expired(cache_item) && locale != LocaleUtil.get_source_locale # local source never expires.
-        refresh_cache(component, locale) # refresh in background
-      end
+      cache_item = load_bundle(component, locale)
       cache_item&.dig(:items)
     end
 
-    def self.load(component, locale)
-      env = SgtnClient::Config.default_environment
-      mode = SgtnClient::Config.configurations[env]['bundle_mode']
-      SgtnClient.logger.debug "[Translation][load]mode=#{mode}"
-      if mode == 'offline'
-        load_o(component, locale)
-      else
-        load_s(component, locale)
-      end
+    def self.load_bundle(component, locale)
+      init_translations unless initialized?
+      super
     end
 
-    def self.load_o(component, locale)
-      env = SgtnClient::Config.default_environment
-      product_name = SgtnClient::Config.configurations[env]['product_name']
-      version = SgtnClient::Config.configurations[env]['version'].to_s
-      translation_bundle = SgtnClient::Config.configurations[env]['translation_bundle']
-      bundlepath = translation_bundle + '/' + product_name + '/' + version + '/' + component + '/messages_' + locale + '.json'
-      SgtnClient::FileUtil.read_json(bundlepath)
-    end
-
-    def self.load_s(component, locale)
-      env = SgtnClient::Config.default_environment
-      product_name = SgtnClient::Config.configurations[env]['product_name']
-      vip_server = SgtnClient::Config.configurations[env]['vip_server']
-      version = SgtnClient::Config.configurations[env]['version'].to_s
-      url = vip_server + '/i18n/api/v2/translation/products/' + product_name + '/versions/' + version + '/locales/' + locale + '/components/' + component + '?checkTranslationStatus=false&machineTranslation=false&pseudo=false'
-      begin
-        obj = SgtnClient::Core::Request.get(url)
-      rescue StandardError => e
-        SgtnClient.logger.error e.message
+    class << self
+      def initialized?
+        @initialized ||= false
       end
-      obj && obj['data']
-    end
 
-    def self.fetch(component, locale)
-      # source locale always uses source bundles
-      return Source.getBundle(component) if LocaleUtil.is_source_locale(locale)
+      def init_translations
+        # TODO: Lock to initialize?
+        env = SgtnClient::Config.default_environment
+        mode = SgtnClient::Config.configurations[env]['bundle_mode']
+        SgtnClient.logger.debug "[Translation][init_translations]mode=#{mode}"
 
-      translation_bundle_thread = Thread.new { load(component, locale) }
-      old_source_bundle = load(component, LocaleUtil.get_source_locale)
-      source_bundle = get_cs(component, LocaleUtil.get_source_locale)
-      translation_bundle = translation_bundle_thread.value
+        if mode == 'offline'
```

Comment:
How do you plan to support mixed mode (falling back to offline bundle if online bundle is not available)?

Reply:
I have a rough idea now. Will create another PR for it.